File: src/features/typing/data/go/maximum_product_subarray.go
package main

func maxProduct(nums []int) int {
	best, maxP, minP := nums[0], nums[0], nums[0]
	for _, num := range nums[1:] {
		if num < 0 {
			maxP, minP = minP, maxP
		}
		maxP = max(num, maxP*num)
		minP = min(num, minP*num)
		if maxP > best {
			best = maxP
		}
	}
	return best
}

File: src/features/typing/data/ts/maximum_product_subarray.ts
function maxProduct(nums: number[]): number {
    let best = nums[0];
    let maxP = nums[0];
    let minP = nums[0];
    for (let i = 1; i < nums.length; i++) {
        const num = nums[i];
        if (num < 0) {
            [maxP, minP] = [minP, maxP];
        }
        maxP = Math.max(num, maxP * num);
        minP = Math.min(num, minP * num);
        if (maxP > best) {
            best = maxP;
        }
    }
    return best;
}
